esque/clients/consumer.py

```diff
@@ -1,49 +1,75 @@
-from contextlib import ExitStack
-from typing import Optional, Tuple
+from abc import ABC
+from abc import abstractmethod
+from typing import Optional, Dict
+from typing import Tuple
 
 import confluent_kafka
 import pendulum
-from confluent_kafka.cimpl import Message, TopicPartition
+from confluent_kafka.cimpl import Message
+from confluent_kafka.cimpl import TopicPartition
 
 from esque.clients.schemaregistry import SchemaRegistryClient
 from esque.config import Config
-from esque.errors import (
-    MessageEmptyException,
-    raise_for_kafka_error,
-    raise_for_message,
-    translate_third_party_exceptions,
-)
-from esque.messages.avromessage import AvroFileWriter
-from esque.messages.message import FileWriter, PlainTextFileWriter
+from esque.errors import EndOfPartitionReachedException
+from esque.errors import MessageEmptyException
+from esque.errors import raise_for_kafka_error
+from esque.errors import raise_for_message
+from esque.errors import translate_third_party_exceptions
+from esque.messages.avromessage import AvroFileWriter, StdOutAvroWriter
+from esque.messages.message import FileWriter, StdOutWriter, GenericWriter
+from esque.messages.message import PlainTextFileWriter
 from esque.ruleparser.ruleengine import RuleTree
 
 
 class AbstractConsumer(ABC):
     def __init__(self, group_id: str, topic_name: str, last: bool, match: str = None):
+        self._match = match
+        self._last = last
+        self._group_id = group_id
+        self._consumer = None
+        self._config = Config().create_confluent_config()
+        if self._match is not None:
+            self._rule_tree = RuleTree(match)
+        else:
+            self._rule_tree = None
+        self._topic_name = topic_name
+        self._use_single_directory_for_messages = False
+        self.writers: Dict[int, GenericWriter] = {}
+
+    @translate_third_party_exceptions
+    def _setup_config(self):
         offset_reset = "earliest"
-        if last:
+        if self._last:
             offset_reset = "latest"
-        if match is not None:
-            self.rule_tree = RuleTree(match)
-
-        self._config = Config().create_confluent_config()
         self._config.update(
             {
-                "group.id": group_id,
+                "group.id": self._group_id,
                 "error_cb": raise_for_kafka_error,
                 # We need to commit offsets manually once we"re sure it got saved
                 # to the sink
                 "enable.auto.commit": True,
-                "enable.partition.eof": False,
+                "enable.partition.eof": True,
                 # We need this to start at the last committed offset instead of the
                 # latest when subscribing for the first time
                 "default.topic.config": {"auto.offset.reset": offset_reset},
             }
         )
-        self._consumer = confluent_kafka.Consumer(self._config)
+
+    @abstractmethod
+    def create_internal_consumer(self):
+        pass
```

Comment:
```suggestion
        raise NotImplementedError()
```
This is the python way of saying: "Please implement this method in your subclass"